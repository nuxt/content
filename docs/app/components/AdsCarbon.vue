<script setup lang="ts">
const el = ref<HTMLDivElement | null>(null)

onMounted(() => {
  if (!el.value) {
    return
  }
  if (document.querySelector('#_carbonads_js')) {
    return
  }

  const script = document.createElement('script')
  script.setAttribute('type', 'text/javascript')
  script.setAttribute('src', 'https://cdn.carbonads.com/carbon.js?serve=CWYIPK7W&placement=contentnuxtcom')
  script.setAttribute('id', '_carbonads_js')

  el.value?.appendChild(script)
})
</script>

<template>
  <div
    ref="el"
    class="carbon"
  />
</template>

<style scoped>
@reference "../assets/css/main.css";

.carbon :deep(#carbonads) {
  @apply relative bg-muted rounded-md hover:bg-elevated w-full transition-colors min-h-[220px] p-2.5;

  .carbon-img {
    @apply flex justify-center w-full;

    & > img {
      @apply !max-w-full w-full rounded-sm;
    }
  }

  .carbon-text {
    @apply text-[13px] transition-colors font-medium text-pretty flex pt-2;
  }

  .carbon-poweredby {
    @apply block text-xs text-muted pt-2;
  }

  &:hover {
    .carbon-text {
      @apply text-highlighted;
    }
  }
}
</style>
