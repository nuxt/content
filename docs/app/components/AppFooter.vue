<script setup lang="ts">
const route = useRoute()

// const items = [{
//   label: 'Figma Kit',
//   to: 'https://www.figma.com/community/file/1288455405058138934',
//   target: '_blank'
// }, {
//   label: 'Playground',
//   to: 'https://stackblitz.com/edit/nuxt-ui',
//   target: '_blank'
// }, {
//   label: 'Roadmap',
//   to: '/roadmap'
// }, {
//   label: 'Releases',
//   to: '/releases'
// }]
</script>

<template>
  <USeparator
    icon="i-simple-icons-nuxtdotjs"
    class="h-px"
  />

  <UFooter>
    <template #left>
      <NuxtLink
        v-if="route.path.startsWith('/pro')"
        to="https://content.nuxt.com/pro/purchase"
        target="_blank"
        class="text-sm text-[--ui-text-muted]"
      >
        Purchase <span class="text-[--ui-text-highlighted]">Nuxt Content Pro</span>
      </NuxtLink>
      <NuxtLink
        v-else
        to="https://github.com/nuxt/content"
        target="_blank"
        class="text-sm text-[--ui-text-muted]"
      >
        Published under <span class="text-[--ui-text-highlighted]">MIT License</span>
      </NuxtLink>
    </template>

    <!-- <UNavigationMenu :items="items" variant="link" color="neutral" /> -->

    <template #right>
      <UButton
        aria-label="Nuxt Content on Discord"
        icon="i-simple-icons-discord"
        to="https://discord.gg/sBXDm6e8SP"
        target="_blank"
        color="neutral"
        variant="ghost"
      />
      <UButton
        aria-label="Nuxt on BlueSKy"
        icon="i-simple-icons-bluesky"
        to="https://go.nuxt.com/bluesky"
        target="_blank"
        color="neutral"
        variant="ghost"
      />
      <UButton
        aria-label="Nuxt on X"
        icon="i-simple-icons-x"
        to="https://x.com/nuxtstudio"
        target="_blank"
        color="neutral"
        variant="ghost"
      />
      <UButton
        aria-label="Nuxt Content on GitHub"
        icon="i-simple-icons-github"
        to="https://github.com/nuxt/content"
        target="_blank"
        color="neutral"
        variant="ghost"
      />
    </template>
  </UFooter>
</template>
