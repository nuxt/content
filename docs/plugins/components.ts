import { UButton } from '#components'

export default defineNuxtPlugin((nuxtApp) => {
  // Set the button as global to use in Markdown
  nuxtApp.vueApp.component('UButton', UButton)
})
