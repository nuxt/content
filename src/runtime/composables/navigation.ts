import { hash } from 'ohash'
import type { NavItem, QueryBuilder, QueryBuilderParams } from '../types'
import { encodeQueryParams } from '../utils/query'
import { jsonStringify } from '../utils/json'
import type { ContentQueryBuilder } from '../types/query'
import { addPrerenderPath, shouldUseClientDB, withContentBase } from './utils'
import { queryContent } from './query'
import { useContentPreview } from './preview'
import { useRuntimeConfig } from '#imports'

export const fetchContentNavigation = async (queryBuilder?: QueryBuilder | QueryBuilderParams | ContentQueryBuilder): Promise<Array<NavItem>> => {
  const { content } = useRuntimeConfig().public

  // Ensure that queryBuilder is an instance of QueryBuilder
  if (typeof queryBuilder?.params !== 'function') {
    queryBuilder = queryContent(queryBuilder as any)
  }

  // Get query params from queryBuilder instance to ensure default values are applied
  const params = queryBuilder.params()

  const apiPath = content.experimental.stripQueryParameters
    ? withContentBase(`/navigation/${process.dev ? '_' : `${hash(params)}.${content.integrity}`}/${encodeQueryParams(params)}.json`)
    : withContentBase(process.dev ? `/navigation/${hash(params)}` : `/navigation/${hash(params)}.${content.integrity}.json`)

  // Add `prefetch` to `<head>` in production
  if (!process.dev && process.server) {
    addPrerenderPath(apiPath)
  }

  if (shouldUseClientDB()) {
    const generateNavigation = await import('./client-db').then(m => m.generateNavigation)
    return generateNavigation(params)
  }

  const data = await $fetch(apiPath as any, {
    method: 'GET',
    responseType: 'json',
    params: content.experimental.stripQueryParameters
      ? undefined
      : {
          _params: jsonStringify(params),
          previewToken: useContentPreview().getPreviewToken()
        }
  })

  // On SSG, all url are redirected to `404.html` when not found, so we need to check the content type
  // to know if the response is a valid JSON or not
  if (typeof data === 'string' && (data as string).startsWith('<!DOCTYPE html>')) {
    throw new Error('Not found')
  }

  return data
}
