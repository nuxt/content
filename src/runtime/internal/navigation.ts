import type { ContentNavigationItem, PageCollectionItemBase, CollectionQueryBuilder } from '@nuxt/content'
import { pascalCase } from 'scule'

/**
 * Create NavItem array to be consumed from runtime plugin.
 */
export async function generateNavigationTree<T extends PageCollectionItemBase>(queryBuilder: CollectionQueryBuilder<T>, extraFields: Array<keyof T> = []) {
  const collecitonItems = await queryBuilder
    .order('stem', 'ASC')
    .orWhere(group => group
      .where('navigation', '<>', 'false')
      .where('navigation', 'IS NULL'),
    )
    .select('navigation', 'stem', 'path', 'title', 'meta', ...(extraFields || []))
    .all() as unknown as PageCollectionItemBase[]

  const { contents, configs } = collecitonItems.reduce((acc, c) => {
    if (String(c.stem).split('/').pop() === '.navigation') {
      c.title = c.title?.toLowerCase() === 'navigation' ? '' : c.title
      const key = c.path!.split('/').slice(0, -1).join('/') || '/'
      acc.configs[key] = {
        ...c,
        ...c.body,
      }
    }
    else {
      acc.contents.push(c)
    }
    return acc
  }, { configs: {}, contents: [] } as { configs: Record<string, PageCollectionItemBase>, contents: PageCollectionItemBase[] })

  // Navigation fields picker
  const pickConfigNavigationFields = (content: PageCollectionItemBase) => ({
    ...pick(['title', ...(extraFields as string[])])(content as unknown as Record<string, unknown>),
    ...content.meta as unknown as Record<string, unknown>,
    ...(isObject(content?.navigation) ? (content.navigation as Record<string, unknown>) : {}),
  })
  const pickNavigationFields = (content: PageCollectionItemBase) => ({
    ...pick(['title', ...(extraFields as string[])])(content as unknown as Record<string, unknown>),
    ...(isObject(content?.navigation) ? (content.navigation as Record<string, unknown>) : {}),
  })

  // Create navigation object
  const nav = contents
    .reduce((nav, content) => {
      // Resolve path and id parts
      const parts = content.path!.substring(1).split('/')
      const idParts = content.stem.split('/')

      // Check if node is `*:index.md`
      const isIndex = !!idParts[idParts.length - 1]?.match(/([1-9]\d*\.)?index/g)

      const getNavItem = (content: PageCollectionItemBase) => ({
        title: content.title,
        path: content.path,
        stem: content.stem,
        children: [],
        ...pickNavigationFields(content),
      } as ContentNavigationItem)

      // Create navigation item from content
      const navItem: ContentNavigationItem = getNavItem(content)

      // Push index
      if (isIndex) {
        // Grab index directory config
        const dirConfig = configs[navItem.path]

        // Drop item if current directory config has `navigation: false`
        if (typeof dirConfig?.navigation !== 'undefined' && dirConfig?.navigation === false) {
          return nav
        }

        // Skip root `index.md` as it has to be pushed as a page
        if (content.path !== '/') {
          const indexItem = getNavItem(content)
          navItem.children!.push(indexItem)
        }

        if (dirConfig) {
          // Merge navigation fields with navItem
          Object.assign(
            navItem,
            pickConfigNavigationFields(dirConfig),
          )
        }
      }

      // First-level item, push it straight to nav
      if (parts.length === 1) {
        // Check for duplicate link
        const existed = nav.find(item => item.path === navItem.path && item.page === false)
        if (isIndex && existed) {
          Object.assign(existed, {
            page: undefined,
            children: [
              ...navItem.children,
              ...existed.children,
            ],
          })
        }
        else {
          nav.push(navItem)
        }

        return nav
      }

      // Find siblings of current item and push them to parent children key
      const siblings = parts.slice(0, -1).reduce((nodes: ContentNavigationItem[], part: string, i: number) => {
        // Part of current path
        const currentPathPart: string = '/' + parts.slice(0, i + 1).join('/')

        // Get current node _dir.yml config
        const conf = configs[currentPathPart]

        // Drop childrens if .navigation.yml has `navigation: false`
        if (typeof conf?.navigation !== 'undefined' && conf.navigation === false) {
          return []
        }

        // Find parent node
        let parent: ContentNavigationItem | undefined = nodes.find(n => n.path === currentPathPart)

        // Create dummy parent if not found
        if (!parent) {
          const navigationConfig = (conf ? pickConfigNavigationFields(conf) : {}) as ContentNavigationItem
          parent = {
            ...navigationConfig,
            title: navigationConfig.title || generateTitle(part),
            path: currentPathPart,
            stem: idParts.slice(0, i + 1).join('/'),
            children: [],
            page: false,
          }
          nodes.push(parent!)
        }

        return parent!.children!
      }, nav)

      // Check for duplicate link
      const existed = siblings.find(item => item.path === navItem.path && item.page === false)
      if (existed) {
        Object.assign(existed, {
          ...navItem,
          page: undefined,
          children: [
            ...navItem.children,
            ...existed.children,
          ],
        })
      }
      else {
        siblings.push(navItem)
      }

      return nav
    }, [] as ContentNavigationItem[])

  return sortAndClear(nav)
}

const collator = new Intl.Collator(undefined, { numeric: true, sensitivity: 'base' })

/**
 * Sort items by path and clear empty children keys.
 */
function sortAndClear(nav: ContentNavigationItem[]) {
  const sorted = nav.sort((a, b) => collator.compare(a.stem!, b.stem!))

  for (const item of sorted) {
    if (item.children?.length) {
      // Sort children
      sortAndClear(item.children)
    }
    else {
      // Remove empty children
      delete item.children
    }
    // Remove path after sort
    // delete item.stem
  }

  return nav
}

/**
 * Returns a new object with the specified keys
 */
function pick(keys?: string[]) {
  return (obj: Record<string, unknown>) => {
    obj = obj || {}
    if (keys && keys.length) {
      return keys
        .filter(key => typeof obj[key] !== 'undefined')
        .reduce((newObj, key) => Object.assign(newObj, { [key]: obj[key] }), {})
    }
    return obj
  }
}

function isObject(obj: unknown) {
  return obj !== null && Object.prototype.toString.call(obj) === '[object Object]'
}

export const generateTitle = (path: string) => path.split(/[\s-]/g).map(pascalCase).join(' ')
