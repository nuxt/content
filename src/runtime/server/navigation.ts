import type { NavItem, ParsedContentMeta } from '../types'
import { generateTitle } from '../transformers/path-meta'
import { useRuntimeConfig } from '#imports'

type PrivateNavItem = NavItem & { path?: string }

/**
 * Create NavItem array to be consumed from runtime plugin.
 */
export function createNav (contents: ParsedContentMeta[], configs: Record<string, ParsedContentMeta>) {
  // Get navigation config from runtimeConfig
  const { navigation } = useRuntimeConfig().public.content

  if (navigation === false) {
    return []
  }

  // Navigation fields picker
  const pickNavigationFields = (content: ParsedContentMeta) => ({
    ...pick(['title', ...navigation.fields])(content),
    ...(isObject(content?.navigation) ? content.navigation : {})
  })

  // Create navigation object
  const nav = contents
    .sort((a, b) => a._path!.localeCompare(b._path!))
    .reduce((nav, content) => {
      // Resolve path and id parts
      const parts = content._path!.substring(1).split('/')
      const idParts = content._id.split(':').slice(1)

      // Check if node is `*:index.md`
      const isIndex = !!idParts[idParts.length - 1].match(/([1-9][0-9]*\.)?index.md/g)

      const getNavItem = (content: ParsedContentMeta) => ({
        title: content.title,
        _path: content._path,
        _file: content._file,
        children: [],
        ...pickNavigationFields(content),
        ...(content._draft ? { _draft: true } : {})
      })

      // Create navigation item from ParsedContent
      const navItem: PrivateNavItem = getNavItem(content)

      // Push index
      if (isIndex) {
        // Grab index directory config
        const dirConfig = configs[navItem._path]

        // Drop item if current directory config has `navigation: false`
        if (typeof dirConfig?.navigation !== 'undefined' && !dirConfig?.navigation) {
          return nav
        }

        // Skip root `index.md` as it has to be pushed as a page
        if (content._path !== '/') {
          const indexItem = getNavItem(content)
          navItem.children!.push(indexItem)
        }

        // Merge navigation fields with navItem
        Object.assign(
          navItem,
          pickNavigationFields(dirConfig)
        )
      }

      // First-level item, push it straight to nav
      if (parts.length === 1) {
        nav.push(navItem)
        return nav
      }

      // Find siblings of current item and push them to parent children key
      const siblings = parts.slice(0, -1).reduce((nodes, part, i) => {
        // Part of current path
        const currentPathPart: string = '/' + parts.slice(0, i + 1).join('/')

        // Get current node _dir.yml config
        const conf = configs[currentPathPart]

        // Drop childrens if _dir.yml has `navigation: false`
        if (typeof conf?.navigation !== 'undefined' && !conf.navigation) {
          return []
        }

        // Find parent node
        let parent: PrivateNavItem | undefined = nodes.find(n => n._path === currentPathPart)

        // Create dummy parent if not found
        if (!parent) {
          parent = {
            title: generateTitle(part),
            _path: currentPathPart,
            _file: content._file,
            children: [],
            ...pickNavigationFields(conf)
          }
          nodes.push(parent!)
        }

        return parent!.children!
      }, nav)

      siblings.push(navItem)

      return nav
    }, [] as PrivateNavItem[])

  return sortAndClear(nav)
}

const collator = new Intl.Collator(undefined, { numeric: true, sensitivity: 'base' })

/**
 * Sort items by path and clear empty children keys.
 */
function sortAndClear (nav: PrivateNavItem[]) {
  nav.forEach((item) => {
    // Remove extension from path
    item._file = item._file.split('.').slice(0, -1).join('.')
  })
  const sorted = nav.sort((a, b) => collator.compare(a._file, b._file))

  for (const item of sorted) {
    if (item.children?.length) {
      // Sort children
      sortAndClear(item.children)
    } else {
      // Remove empty children
      delete item.children
    }
    // Remove path after sort
    delete item._file
  }

  return nav
}

/**
 * Returns a new object with the specified keys
 **/
function pick (keys?: string[]) {
  return (obj: any) => {
    obj = obj || {}
    if (keys && keys.length) {
      return keys
        .filter(key => typeof obj[key] !== 'undefined')
        .reduce((newObj, key) => Object.assign(newObj, { [key]: obj[key] }), {})
    }
    return obj
  }
}

function isObject (obj: any) {
  return Object.prototype.toString.call(obj) === '[object Object]'
}
