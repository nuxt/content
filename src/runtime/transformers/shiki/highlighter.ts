import { getHighlighter, BUNDLED_LANGUAGES, BUNDLED_THEMES, Lang, Theme as ShikiTheme, Highlighter } from 'shiki-es'
import consola from 'consola'
import type { ModuleOptions } from '../../../module'
import { createSingleton } from '../utils'
import mdcTMLanguage from './languages/mdc.tmLanguage'
import type { MarkdownNode, HighlighterOptions, Theme, HighlightThemedToken, HighlightThemedTokenLine, TokenColorMap } from './types'

// Re-create logger locally as utils cannot be imported from here
const logger = consola.withScope('@nuxt/content')

/**
 * Resolve Shiki compatible lang from string.
 *
 * Used to resolve lang from both languages id's and aliases.
 */
const resolveLang = (lang: string) =>
  (BUNDLED_LANGUAGES.find(l => l.id === lang || l.aliases?.includes(lang)))

/**
 * Resolve Shiki compatible theme from string.
 */
const resolveTheme = (theme: string | Record<string, string>): Record<string, ShikiTheme> | undefined => {
  if (!theme) {
    return
  }
  if (typeof theme === 'string') {
    theme = {
      default: theme
    }
  }

  return Object.entries(theme).reduce((acc, [key, value]) => {
    acc[key] = BUNDLED_THEMES.find(t => t === value)!
    return acc
  }, {} as Record<string, ShikiTheme>)
}

export const useShikiHighlighter = createSingleton((opts?: Exclude<ModuleOptions['highlight'], false>) => {
  // Grab highlighter config from publicRuntimeConfig
  const { theme, preload } = opts || {}

  let promise: Promise<Highlighter> | undefined
  const getShikiHighlighter = () => {
    if (!promise) {
      // Initialize highlighter with defaults
      promise = getHighlighter({
        theme: (theme as any)?.default || theme || 'dark-plus',
        langs: [
          ...(preload || []),
          'diff',
          'json',
          'js',
          'ts',
          'css',
          'shell',
          'html',
          'md',
          'yaml',
          'vue',
          {
            id: 'md',
            scopeName: 'text.markdown.mdc',
            path: 'mdc.tmLanguage.json',
            aliases: ['markdown', 'md', 'mdc'],
            grammar: mdcTMLanguage
          }
        ] as any[]
      }).then((highlighter) => {
        // Load all themes on-demand
        const themes = Object.values(typeof theme === 'string' ? { default: theme } : (theme || {}))

        if (themes.length) {
          return Promise
            .all(themes.map(theme => highlighter.loadTheme(theme)))
            .then(() => highlighter)
        }
        return highlighter
      })
    }
    return promise
  }

  const getHighlightedTokens = async (code: string, lang: Lang, theme: Theme) => {
    const highlighter = await getShikiHighlighter()
    // Remove trailing carriage returns
    code = code.replace(/\n+$/, '')
    // Resolve lang & theme (i.e check if shiki supports them)
    lang = (resolveLang(lang || '')?.id || lang) as Lang
    theme = resolveTheme(theme || '') || { default: highlighter.getTheme() as any as ShikiTheme }

    // Skip highlight if lang is not supported
    if (!lang) {
      return [[{ content: code }]]
    }

    // Load supported language on-demand
    if (!highlighter.getLoadedLanguages().includes(lang)) {
      const languageRegistration = resolveLang(lang)

      if (languageRegistration) {
        await highlighter.loadLanguage(languageRegistration)
      } else {
        logger.warn(`Language '${lang}' is not supported by shiki. Skipping highlight.`)
        return [[{ content: code }]]
      }
    }

    // Load supported theme on-demand
    const newThemes = Object.values(theme).filter(t => !highlighter.getLoadedThemes().includes(t))
    if (newThemes.length) {
      await Promise.all(newThemes.map(highlighter.loadTheme))
    }

    // Highlight code
    const coloredTokens = Object.entries(theme).map(([key, theme]) => {
      const tokens = highlighter.codeToThemedTokens(code, lang, theme, { includeExplanation: false })
      return {
        key,
        theme,
        tokens
      }
    })

    const highlightedCode: HighlightThemedToken[][] = []
    for (const line in coloredTokens[0].tokens) {
      highlightedCode[line] = coloredTokens.reduce((acc, color) => {
        return mergeLines({
          key: coloredTokens[0].key,
          tokens: acc
        }, {
          key: color.key,
          tokens: color.tokens[line]
        })
      }, coloredTokens[0].tokens[line] as HighlightThemedToken[])
    }

    return highlightedCode
  }

  const getHighlightedAST = async (code: string, lang: Lang, theme: Theme, opts?: Partial<HighlighterOptions>): Promise<Array<MarkdownNode>> => {
    const lines = await getHighlightedTokens(code, lang, theme)
    const { highlights = [], colorMap = {} } = opts || {}

    return lines.map((line, lineIndex) => ({
      type: 'element',
      tag: 'span',
      props: { class: ['line', highlights.includes(lineIndex + 1) ? 'highlight' : ''].join(' ').trim() },
      children: line.map(tokenSpan)
    }))

    function getColorProps (token: { color?: string | object }) {
      if (!token.color) {
        return {}
      }
      if (typeof token.color === 'string') {
        return { style: { color: token.color } }
      }
      const key = Object.values(token.color).join('')
      if (!colorMap[key]) {
        colorMap[key] = {
          colors: token.color,
          className: 'ct-' + Math.random().toString(16).substring(2, 8) // hash(key)
        }
      }
      return { class: colorMap[key].className }
    }

    function tokenSpan (token: { content: string, color?: string | object }) {
      return {
        type: 'element',
        tag: 'span',
        props: getColorProps(token),
        children: [{ type: 'text', value: token.content }]
      }
    }
  }

  const getHighlightedCode = async (code: string, lang: Lang, theme: Theme, opts?: Partial<HighlighterOptions>) => {
    const colorMap = opts?.colorMap || {}
    const highlights = opts?.highlights || []
    const ast = await getHighlightedAST(code, lang, theme, { colorMap, highlights })

    function renderNode (node: any) {
      if (node.type === 'text') {
        return node.value.replace(/</g, '&lt;').replace(/>/g, '&gt;')
      }
      const children = node.children.map(renderNode).join('')
      return `<${node.tag} class="${node.props.class}">${children}</${node.tag}>`
    }

    return {
      code: ast.map(renderNode).join(''),
      styles: generateStyles(colorMap)
    }
  }

  const generateStyles = (colorMap: TokenColorMap) => {
    const colors: string[] = []
    for (const colorClass of Object.values(colorMap)) {
      Object.entries(colorClass.colors).forEach(([variant, color]) => {
        if (variant === 'default') {
          colors.unshift(`.${colorClass.className}{color:${color}}`)
        } else {
          colors.push(`.${variant} .${colorClass.className}{color:${color}}`)
        }
      })
    }
    return colors.join('\n')
  }

  return {
    getHighlightedTokens,
    getHighlightedAST,
    getHighlightedCode,
    generateStyles
  }
})

function mergeLines (line1: HighlightThemedTokenLine, line2: HighlightThemedTokenLine) {
  const mergedTokens: HighlightThemedToken[] = []
  const getColors = (h: HighlightThemedTokenLine, i: number) => typeof h.tokens[i].color === 'string' ? { [h.key]: h.tokens[i].color } : h.tokens[i].color as object

  const right = {
    key: line1.key,
    tokens: line1.tokens.slice()
  }
  const left = {
    key: line2.key,
    tokens: line2.tokens.slice()
  }
  let index = 0
  while (index < right.tokens.length) {
    const rightToken = right.tokens[index]
    const leftToken = left.tokens[index]

    if (rightToken.content === leftToken.content) {
      mergedTokens.push({
        content: rightToken.content,
        color: {
          ...getColors(right, index),
          ...getColors(left, index)
        }
      })
      index += 1
      continue
    }

    if (rightToken.content.startsWith(leftToken.content)) {
      const nextRightToken = {
        ...rightToken,
        content: rightToken.content.slice(leftToken.content.length)
      }
      rightToken.content = leftToken.content
      right.tokens.splice(index + 1, 0, nextRightToken)
      continue
    }

    if (leftToken.content.startsWith(rightToken.content)) {
      const nextLeftToken = {
        ...leftToken,
        content: leftToken.content.slice(rightToken.content.length)
      }
      leftToken.content = rightToken.content
      left.tokens.splice(index + 1, 0, nextLeftToken)
      continue
    }

    throw new Error('Unexpected token')
  }
  return mergedTokens
}
