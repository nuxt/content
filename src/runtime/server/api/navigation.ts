import { defineEventHandler } from 'h3'
import { cacheStorage, serverQueryContent } from '../storage'
import { createNav } from '../navigation'
import { ParsedContent, ParsedContentMeta } from '../../types'
import { getContentQuery } from '../../utils/query'
import { isPreview } from '../preview'

export default defineEventHandler(async (event) => {
  const query = getContentQuery(event)

  // Read from cache if not preview and there is no query
  if (!isPreview(event) && Object.keys(query).length === 0) {
    const cache = await cacheStorage.getItem('content-navigation.json')
    if (cache) {
      return cache
    }
  }

  const contents = await serverQueryContent(event, query)
    .where({
      /**
       * Partial contents are not included in the navigation
       * A partial content is a content that has `_` prefix in its path
       */
      _partial: false,
      /**
       * Exclude any pages which have opted out of navigation via frontmatter.
       */
      navigation: {
        $ne: false
      }
    })
    .find()

  const dirConfigs = await serverQueryContent(event).where({ _path: /\/_dir$/i, _partial: true }).find()

  const configs = dirConfigs.reduce((configs, conf: ParsedContent) => {
    if (conf.title?.toLowerCase() === 'dir') {
      conf.title = undefined
    }
    const key = conf._path!.split('/').slice(0, -1).join('/') || '/'
    configs[key] = {
      ...conf,
      // Extract meta from body. (non MD files)
      ...conf.body
    }
    return configs
  }, {} as Record<string, ParsedContentMeta>)

  return createNav(contents as ParsedContentMeta[], configs)
})
