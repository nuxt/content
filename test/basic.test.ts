import { fileURLToPath } from 'url'
import { test, describe, expect } from 'vitest'
import { setup, $fetch } from '@nuxt/test-utils'
import { testMarkdownParser } from './features/parser-markdown'
import { testPathMetaTransformer } from './features/transformer-path-meta'
import { testYamlParser } from './features/parser-yaml'
import { testNavigation } from './features/navigation'
// import { testMDCComponent } from './features/mdc-component'
import { testJSONParser } from './features/parser-json'
import { testCSVParser } from './features/parser-csv'
import { testRegex } from './features/regex'
import { testMarkdownParserExcerpt } from './features/parser-markdown-excerpt'
import { testParserHooks } from './features/parser-hooks'
import { testModuleOptions } from './features/module-options'
import { testContentQuery } from './features/content-query'
import { testHighlighter } from './features/highlighter'
import { testMarkdownRenderer } from './features/renderer-markdown'
import { testParserOptions } from './features/parser-options'
import { testComponents } from './features/components'
import { testLocales } from './features/locales'
import { testIgnores } from './features/ignores'

// const spyConsoleWarn = vi.spyOn(global.console, 'warn')

describe('Basic usage', async () => {
  await setup({
    rootDir: fileURLToPath(new URL('./fixtures/basic', import.meta.url)),
    server: true
  })

  test('Multi part path', async () => {
    const html = await $fetch('/features/multi-part-path')
    expect(html).contains('Persian')
  })

  test('Japanese path', async () => {
    const html = await $fetch('/' + encodeURIComponent('こんにちは'))
    expect(html).contains('🎨 こんにちは')
  })

  test('Partials specials chars', async () => {
    const html = await $fetch('/_partial/content-(v2)')
    expect(html).contains('Content (v2)')
  })

  test('Partials specials chars', async () => {
    const html = await $fetch('/_partial/markdown')
    expect(html).contains('><!--[--> Default title <!--]--></h1>')
    expect(html).contains('<p><!--[-->p1<!--]--></p>')
  })

  test('Fetch only title and excerpt', async () => {
    const html = await $fetch('/excerpt-only')
    expect(html).contains('Excerpt paragraph')
    expect(html).not.contains('Rest of the content')
  })

  // test('Warning for invalid file name', () => {
  //   expect(spyConsoleWarn).toHaveBeenCalled()
  //   expect(spyConsoleWarn).toHaveBeenCalledWith('Ignoring [content:with-\'invalid\'-char.md]. File name should not contain any of the following characters: \', ", ?, #, /')
  // })

  testLocales()

  testComponents()

  testContentQuery()

  testNavigation()

  testMarkdownParser()

  testMarkdownRenderer()

  testMarkdownParserExcerpt()

  testYamlParser()

  testCSVParser()

  testJSONParser()

  testPathMetaTransformer()

  // testMDCComponent()

  testRegex()

  testParserHooks()

  testModuleOptions()

  testHighlighter()

  testParserOptions()

  testIgnores()
})
