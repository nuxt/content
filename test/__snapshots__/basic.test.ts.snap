// Vitest Snapshot v1

exports[`Basic usage > Content Queries > Get contents index > basic-index-body 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "Index",
        },
      ],
      "props": {
        "id": "index",
      },
      "tag": "h1",
      "type": "element",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "Hello World",
        },
      ],
      "props": {},
      "tag": "p",
      "type": "element",
    },
  ],
  "toc": {
    "depth": 2,
    "links": [],
    "searchDepth": 2,
    "title": "",
  },
  "type": "root",
}
`;

exports[`Basic usage > Get contents index > basic-index-body 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "Index",
        },
      ],
      "props": {
        "id": "index",
      },
      "tag": "h1",
      "type": "element",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "Hello World",
        },
      ],
      "props": {},
      "tag": "p",
      "type": "element",
    },
  ],
  "toc": {
    "depth": 2,
    "links": [],
    "searchDepth": 2,
    "title": "",
  },
  "type": "root",
}
`;

exports[`Basic usage > Navigation > Get cats navigation > basic-navigation-cats 1`] = `
[
  {
    "_path": "/cats",
    "children": [
      {
        "_path": "/cats/bombay",
        "title": "Bombay",
      },
      {
        "_path": "/cats",
        "title": "Cats",
      },
      {
        "_path": "/cats/persian",
        "title": "Persian",
      },
      {
        "_path": "/cats/ragdoll",
        "title": "Ragdoll",
      },
    ],
    "title": "Cats list",
  },
]
`;

exports[`Basic usage > Navigation > Get dogs navigation > basic-navigation-dogs 1`] = `
[
  {
    "_path": "/dogs",
    "children": [
      {
        "_path": "/dogs/bulldog",
        "title": "Bulldog",
      },
      {
        "_path": "/dogs/german-shepherd",
        "title": "German Shepherd",
      },
    ],
    "icon": "üê∂",
    "title": "Dogs List",
  },
]
`;

exports[`fixtures:basic > Get contents index > basic-index-body 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "Index",
        },
      ],
      "props": {
        "id": "index",
      },
      "tag": "h1",
      "type": "element",
    },
  ],
  "toc": {
    "depth": 2,
    "links": [],
    "searchDepth": 2,
    "title": "",
  },
  "type": "root",
}
`;

exports[`fixtures:basic > navigation > _dir.yml should be able to filter navigation tree > basic-navigation-dogs 1`] = `
[
  {
    "_path": "/navigation",
    "children": [
      {
        "_path": "/navigation/hidden-dir",
        "children": [
          {
            "_path": "/navigation/hidden-dir",
            "title": "Hello World!",
          },
        ],
        "title": "Hello World!",
      },
      {
        "_path": "/navigation",
        "title": "Hello World",
      },
    ],
    "title": "Hello World",
  },
]
`;

exports[`fixtures:basic > navigation > Get cats navigation > basic-navigation-cats 1`] = `
[
  {
    "_path": "/cats",
    "children": [
      {
        "_path": "/cats/bombay",
        "title": "Bombay",
      },
      {
        "_path": "/cats",
        "title": "Cats",
      },
      {
        "_path": "/cats/persian",
        "title": "Persian",
      },
      {
        "_path": "/cats/ragdoll",
        "title": "Ragdoll",
      },
    ],
    "title": "Cats list",
  },
]
`;

exports[`fixtures:basic > navigation > Get dogs navigation > basic-navigation-dogs 1`] = `
[
  {
    "_path": "/dogs",
    "children": [
      {
        "_path": "/dogs/bulldog",
        "title": "Bulldog",
      },
      {
        "_path": "/dogs/german-shepherd",
        "title": "German Shepherd",
      },
    ],
    "icon": "üê∂",
    "title": "Dogs List",
  },
]
`;
